NE-1744: Updating instructions for GettingStarted (#11)
openshift/network-edge-tools: docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md

# Getting Started

In this section, we will show you how to install and configure Gateway API via the Ingress Operator. We will configure a simple HTTPRoute and backend to demo basic Gateway API functionality.

> Note: Since this is still a Developer Preview, it is important to note that that it may not be flawless, and it is possible for bugs to be present.

Reviewer: Since we now have a GA, we'll need to reword alot of this.
```suggestion
Gateway API via Ingress Operator is GA in OpenShift 4.19.  These intstructions were written for versions prior 4.19, when Gateway API was considered a Developer Preview.  When using versions prior to 4.19, it is important to note that it is possible for bugs to be present.
```

## Prerequisites

>   * A non-production OpenShift Cluster, version 4.13.5 or greater, on a cloud platform that supports Kubernetes service load balancers (e.g. GCP, AWS, and Azure)

Reviewer: Let's pick a more recent version, like 4.18.0.
```suggestion
  * Prior to GA 4.19, use a non-production OpenShift Cluster
  * OpenShift version 4.18 or greater, on a cloud platform that supports Kubernetes service load balancers (e.g. GCP, AWS, and Azure)
```

>   * If OSSM is already installed, it must be OSSM 2.4.0 for 4.13

Reviewer: ```suggestion
  * If OSSM is already installed, it must be at least 
  ** OSSM 2.6 for OpenShift 4.18
  ** OSSM 3.0.1 for OpenShift 4.19
  ** OSSM 3.1.0 for OpenShift 4.20
```

>     * Version Note: For OpenShift 4.20, the minimum required OSSM version is 2.6.2.

Reviewer: This can be removed.

>     * Version Note: For OpenShift 4.20, the minimum required OSSM version is 2.6.2.

Reviewer: delete this line

>   * If OSSM is already installed, you must not have any existing ServiceMeshControlPlane CRs

Reviewer: ```suggestion
  * In OpenShift 4.18, if OSSM is already installed, you must not have any existing ServiceMeshControlPlane CRs
```

>   * The [OpenShift CLI](https://docs.openshift.com/container-platform/4.13/cli_reference/openshift_cli/getting-started-cli.html#cli-installing-cli_cli-developer-commands) binary (`oc`)

Reviewer: Update to 4.20 link.

-----

## Installation via the Ingress Operator

To install Gateway API using the Ingress Operator, you will do the following:

> 1.  Grant the Ingress Operator the cluster-admin role (no longer required in 4.20)
> 2.  Enable the feature gate (no longer required to 4.20)
> 3.  Create a GatewayClass with the controller name `openshift.io/gateway-controller`

Reviewer: ```suggestion
1.  For OCP versions before 4.20, grant the Ingress Operator the cluster-admin role, enable the feature gate, and
create a GatewayClass with the controller name `openshift.io/gateway-controller`
2. For OCP 4.20 and later, just create a GatewayClass with the controller name `openshift.io/gateway-controller/v1` (note the additional version suffix)
```

Let’s walk through each of these steps in more detail.

### 1\. Grant Cluster-Admin Role (Version Specific)

First, grant the Ingress Operator the cluster-admin role. This role is required in order for the Ingress Operator to create the SMCP CR for OSSM:

>   * **OpenShift 4.20 and beyond:** This step is no longer required and should be skipped.

Reviewer 2: do we even provide support for OSSM 2.x? do we need this section about SMCP?

Reviewer: This is an old doc being updated for 4.20+, and we still allow people to experiment on previous versions.

```console
$ oc adm policy add-cluster-role-to-user cluster-admin -z ingress-operator -n openshift-ingress-operator
clusterrole.rbac.authorization.k8s.io/cluster-admin added: "ingress-operator"
```

It is anticipated that this need for cluster-admin permissions will be eliminated in a future release of Gateway API in OpenShift.

> ### 2\. Enable the Feature Gate

Reviewer 2: need to remove feature gate stuff

Reviewer: This is still valid for 4.18.

Enable the feature gate for Gateway API. This will instruct the cluster Ingress Operator to install the Gateway API CRDs.
Please note that, since feature gates are stored in the machine config, patching the feature gate config will result in the machine config operator rebooting each node sequentially, according to your machine config pool configuration:

> * **OpenShift 4.20 and beyond:** This step is no longer required and should be skipped.

Reviewer 2: this should be 4.19 and above? It went GA then

```console
$ oc patch featuregates/cluster --type=merge --patch='{"spec":{"featureSet":"CustomNoUpgrade","customNoUpgrade":{"enabled":["GatewayAPI"]}}}'
featuregate.config.openshift.io/cluster patched
```

Wait for the ingress operator to successfully install the Gateway API CRDs:

```console
$ oc get crd gatewayclasses.gateway.networking.k8s.io httproutes.gateway.networking.k8s.io gateways.gateway.networking.k8s.io referencegrants.gateway.networking.k8s.io
NAME                                        CREATED AT
gatewayclasses.gateway.networking.k8s.io    2025-02-04T10:49:54Z
httproutes.gateway.networking.k8s.io        2025-02-04T10:49:55Z

> gateways.gateway.networking.k8s.io          2025-02-04T10:49:54Z

Reviewer: This whole section should be re-written with a Header like: "For OCP version before 4.20".

# Getting Started

In this section, we will show you how to install and configure Gateway API via the Ingress Operator. We will configure a simple HTTPRoute and backend to demo basic Gateway API functionality.

> Note: Gateway API via Ingress Operator is GA in OpenShift 4.19.  These instructions were written for versions prior to 4.19, when Gateway API was considered a Developer Preview.  When using versions prior to 4.19, it is important to note that bugs may be present.

Reviewer 3: This is a confusing statement, especially "bugs may be present".  The examples should have been tested on the Developer Preview versions, and they should still work on the GA version.  If there are any discrepancies, those should be noted in the document (for example, the example GatewayClass should have a note that the Developer Preview and GA use different controller names), or if that is too much effort, just say that the instructions were tested on the Developer Preview and might require some adjustment for GA.

## Prerequisites

  * Prior to GA 4.19, use a non-production OpenShift Cluster
  * OpenShift version 4.18 or greater, on a cloud platform that supports Kubernetes service load balancers (e.g. GCP, AWS, and Azure)

>   * If OSSM is already installed, it must be at least:

Reviewer 3: I'm still not sure that "at least" is correct.

>   * If OSSM is already installed, the minimum version must be:

Reviewer 3: Is it a _minimum_ version, or does the version need to be _exactly_ the one listed here?

    * OSSM 2.6 for OpenShift 4.18
    * OSSM 3.0.1 for OpenShift 4.19
    * OSSM 3.1.0 for OpenShift 4.20
  * In OpenShift 4.18, if OSSM is already installed, you must not have any existing ServiceMeshControlPlane CRs
  * The [OpenShift CLI](https://docs.redhat.com/en/documentation/openshift_container_platform/4.20/html/cli_tools/openshift-cli-oc) binary (`oc`)

-----

## Installation via the Ingress Operator

To install Gateway API using the Ingress Operator:

* **OCP 4.20 and later:** Create a GatewayClass with controller name `openshift.io/gateway-controller/v1` (see [Step 3](#3-create-the-gatewayclass))

> * **OCP versions before 4.20:** Follow Steps 1-3 below (grant cluster-admin role, enable the `GatewayAPI` feature gate, create GatewayClass)

Reviewer 3: OpenShift 4.19 has the controller name with the "/v1" suffix: https://github.com/openshift/cluster-ingress-operator/blob/release-4.19/pkg/operator/controller/names.go#L65

I think you need to fix a number of places that say "4.20 or later" and "before 4.20" to say "4.19 or later" and "before 4.19" instead.

### 1\. Grant Cluster-Admin Role (For OCP version before 4.20)

  * **OpenShift 4.20 and beyond:** This step is no longer required and should be skipped.

First, grant the Ingress Operator the cluster-admin role. This role is required in order for the Ingress Operator to create the SMCP CR for OSSM:

```console
$ oc adm policy add-cluster-role-to-user cluster-admin -z ingress-operator -n openshift-ingress-operator
clusterrole.rbac.authorization.k8s.io/cluster-admin added: "ingress-operator"
```

> It is anticipated that this need for cluster-admin permissions will be eliminated in a future release of Gateway API in OpenShift.

Reviewer: This could be removed, since we already say 

```
OpenShift 4.20 and beyond: This step is no longer required and should be skipped.
```

Reviewer 3: The redundant sentence is still present.

### 2\. Enable the Feature Gate (For OCP version before 4.19)

  * **OpenShift 4.19 and beyond:** This step is no longer required and should be skipped.

Enable the feature gate for Gateway API. This will instruct the cluster Ingress Operator to install the Gateway API CRDs.
Please note that, since feature gates are stored in the machine config, patching the feature gate config will result in the machine config operator rebooting each node sequentially, according to your machine config pool configuration:

```console
$ oc patch featuregates/cluster --type=merge --patch='{"spec":{"featureSet":"CustomNoUpgrade","customNoUpgrade":{"enabled":["GatewayAPI"]}}}'
featuregate.config.openshift.io/cluster patched
```

Wait for the ingress operator to successfully install the Gateway API CRDs:

> > **Note:** OCP 4.20 and later also includes the `grpcroutes.gateway.networking.k8s.io` CRD for routing gRPC traffic.

Reviewer 3: ```suggestion
> **Note:** OCP 4.19 and later also includes the `grpcroutes.gateway.networking.k8s.io` CRD for routing gRPC traffic.
```

gatewayclasses.gateway.networking.k8s.io    2025-02-04T10:49:54Z
httproutes.gateway.networking.k8s.io        2025-02-04T10:49:55Z
gateways.gateway.networking.k8s.io          2025-02-04T10:49:54Z
referencegrants.gateway.networking.k8s.io   2025-02-04T10:49:55Z
```

### 3\. Create the GatewayClass

Next, create the GatewayClass. The controller name differs by version:

  * **Prior to OpenShift 4.19:** Use controller name `openshift.io/gateway-controller`
    ```console
    $ oc create -f -<<'EOF'
    apiVersion: gateway.networking.k8s.io/v1beta1
    kind: GatewayClass
    metadata:
      name: openshift-default
    spec:
      controllerName: openshift.io/gateway-controller
    EOF
    gatewayclass.gateway.networking.k8s.io/openshift-default created
    ```

  * **OpenShift 4.19 and beyond:** Use controller name `openshift.io/gateway-controller/v1`
    ```console
    $ oc create -f -<<'EOF'
    apiVersion: gateway.networking.k8s.io/v1
    kind: GatewayClass
    metadata:
      name: openshift-default
    spec:
      controllerName: openshift.io/gateway-controller/v1
    EOF
    gatewayclass.gateway.networking.k8s.io/openshift-default created
    ```

> Behind the scenes, the Ingress Operator creates a ServiceMeshControlPlane (SMCP) CR, which is the configuration API for OSSM. Wait a minute or two for the SMCP to be created and become ready:

Reviewer 3: This statement about the SMCP belongs under the "Prior to OpenShift 4.19" section.

> Behind the scenes, the Ingress Operator creates a ServiceMeshControlPlane (SMCP) CR, which is the configuration API for OSSM. Wait a minute or two for the SMCP to be created and become ready:

Reviewer: This is not quite correct, but it is corrected later in the text.
```suggestion
Behind the scenes, the Ingress Operator creates either an Istio or ServiceMeshControlPlane (SMCP) CR.  The SMCP CR is the configuration API for OSSM prior to OCP 4.20, and the Istio CR takes its place in 4.20 and later. Wait a minute or two for the Istio/SMCP to be created and become ready:
```

>   * **Prior to OpenShift 4.20:** OSSM uses ServiceMeshControlPlane (SMCP) resources.

Reviewer 3: ```suggestion
  * **Prior to OpenShift 4.19:** OSSM uses ServiceMeshControlPlane (SMCP) resources.
```

    ```console
    $ oc get smcp -n openshift-ingress
    NAME                READY   STATUS            PROFILES      VERSION   AGE
    openshift-gateway   5/5     ComponentsReady   ["default"]   2.4.0     10m
    ```

>   * **OpenShift 4.20 and beyond:** OSSM 3 uses Istio resources instead of SMCP.

Reviewer 3: ```suggestion
  * **OpenShift 4.19 and beyond:** OSSM 3 uses Istio resources instead of SMCP.
```

>   * **OpenShift 4.20 and beyond:** OSSM 3.x uses Istio resources instead of SMCP.

Reviewer 3: ```suggestion
  * **OpenShift 4.20 and beyond:** OSSM 3._y_ uses Istio resources instead of SMCP.
```
or simply
```suggestion
  * **OpenShift 4.20 and beyond:** OSSM 3 uses Istio resources instead of SMCP.
```

    ```console
    $ oc get istio -A
    NAME                NAMESPACE           PROFILE   REVISIONS   READY   IN USE   ACTIVE REVISION     STATUS    VERSION   AGE
    openshift-gateway   openshift-ingress             1           1       0        openshift-gateway   Healthy   v1.26.2   2m
    ```

OSSM will create Istio deployments in the `openshift-ingress` namespace. Ensure the pods for the deployments are running.

>   * **Prior to OpenShift 4.20:** OSSM creates a shared `istio-ingressgateway` deployment and the control plane.

Reviewer 3: ```suggestion
  * **Prior to OpenShift 4.19:** OSSM creates a shared `istio-ingressgateway` deployment and the control plane.
````

    ```console
    $ oc get deployment -n openshift-ingress
    NAME                      READY   UP-TO-DATE   AVAILABLE   AGE
    istio-ingressgateway      1/1     1            1           42s
    istiod-openshift-gateway  1/1     1            1           55s
    router-default            2/2     2            2           6h4m
    ```

>   * **OpenShift 4.20 and beyond:** Only the Istio control plane (`istiod`) is created at this stage. Individual gateway deployments are created when you create Gateway resources.

Reviewer 3: ```suggestion
  * **OpenShift 4.19 and beyond:** Only the Istio control plane (`istiod`) is created at this stage. Individual gateway deployments are created when you create Gateway resources.
````

# Getting Started

In this section, we will show you how to install and configure Gateway API via the Ingress Operator. We will configure a simple HTTPRoute and backend to demo basic Gateway API functionality.

Note: Gateway API via Ingress Operator is GA in OpenShift 4.19.  These instructions were written for versions prior to 4.19, when Gateway API was considered a Developer Preview.  When using versions prior to 4.19, it is important to note that bugs may be present.
## Prerequisites

  * Prior to GA 4.19, use a non-production OpenShift Cluster
  * OpenShift version 4.18 or greater, on a cloud platform that supports Kubernetes service load balancers (e.g. GCP, AWS, and Azure)
  * If OSSM is already installed, the minimum version must be:
    * OSSM 2.6 for OpenShift 4.18
    * OSSM 3.0.1 for OpenShift 4.19
    * OSSM 3.1.0 for OpenShift 4.20
  * In OpenShift 4.18, if OSSM is already installed, you must not have any existing ServiceMeshControlPlane CRs

>   * The [OpenShift CLI](https://docs.okd.io/latest/cli_reference/openshift_cli/getting-started-cli.html#:~:text=using%20an%20RPM.-,Installing%20the%20OpenShift%20CLI,-You%20can%20install) binary (`oc`)

Reviewer 4: shouldn't this point to official docs?

-----

## Installation via the Ingress Operator

To install Gateway API using the Ingress Operator, you will do the following:

> 1.  For OCP versions before 4.20, grant the Ingress Operator the cluster-admin role, enable the feature gate, and

Reviewer 4: which feature gate? should it be part of the text here?

Reviewer 4: ah I can see it below. Does it make sense to remove these bullet points here and just direct users to the details then?

  * **OpenShift 4.20 and beyond:** This step is no longer required and should be skipped.

Enable the feature gate for Gateway API. This will instruct the cluster Ingress Operator to install the Gateway API CRDs.
Please note that, since feature gates are stored in the machine config, patching the feature gate config will result in the machine config operator rebooting each node sequentially, according to your machine config pool configuration:

```console
$ oc patch featuregates/cluster --type=merge --patch='{"spec":{"featureSet":"CustomNoUpgrade","customNoUpgrade":{"enabled":["GatewayAPI"]}}}'
featuregate.config.openshift.io/cluster patched
```

Wait for the ingress operator to successfully install the Gateway API CRDs:

```console
$ oc get crd gatewayclasses.gateway.networking.k8s.io httproutes.gateway.networking.k8s.io gateways.gateway.networking.k8s.io referencegrants.gateway.networking.k8s.io
NAME                                        CREATED AT
gatewayclasses.gateway.networking.k8s.io    2025-02-04T10:49:54Z
httproutes.gateway.networking.k8s.io        2025-02-04T10:49:55Z
gateways.gateway.networking.k8s.io          2025-02-04T10:49:54Z

> referencegrants.gateway.networking.k8s.io   2025-02-04T10:49:55Z

Reviewer 2: 4.20 beyond also supports grpc route crds, do we want to add that in?

Reviewer: It was added.

    ```console
    $ oc create -f -<<'EOF'
    apiVersion: gateway.networking.k8s.io/v1
    kind: GatewayClass
    metadata:
      name: openshift-default
    spec:
      controllerName: openshift.io/gateway-controller/v1
    EOF
    gatewayclass.gateway.networking.k8s.io/openshift-default created
    ```

Behind the scenes, the Ingress Operator creates a ServiceMeshControlPlane (SMCP) CR, which is the configuration API for OSSM. Wait a minute or two for the SMCP to be created and become ready:

  * **Prior to OpenShift 4.20:** OSSM uses ServiceMeshControlPlane (SMCP) resources.
    ```console
    $ oc get smcp -n openshift-ingress
    NAME                READY   STATUS            PROFILES      VERSION   AGE
    openshift-gateway   5/5     ComponentsReady   ["default"]   2.4.0     10m
    ```

  * **OpenShift 4.20 and beyond:** OSSM 3.x uses Istio resources instead of SMCP.
    ```console
    $ oc get istio -A
    NAME                NAMESPACE           PROFILE   REVISIONS   READY   IN USE   ACTIVE REVISION     STATUS    VERSION   AGE
    openshift-gateway   openshift-ingress             1           1       0        openshift-gateway   Healthy   v1.26.2   2m

>     ```

Reviewer 2: isn't istio cluster scoped? `-A` should not be required

OSSM will create Istio deployments in the `openshift-ingress` namespace. Ensure the pods for the deployments are running.

  * **Prior to OpenShift 4.20:** OSSM creates a shared `istio-ingressgateway` deployment and the control plane.
    ```console
    $ oc get deployment -n openshift-ingress
    NAME                      READY   UP-TO-DATE   AVAILABLE   AGE
    istio-ingressgateway      1/1     1            1           42s
    istiod-openshift-gateway  1/1     1            1           55s
    router-default            2/2     2            2           6h4m
    ```

  * **OpenShift 4.20 and beyond:** Only the Istio control plane (`istiod`) is created at this stage. Individual gateway deployments are created when you create Gateway resources.
    ```console
    $ oc get deployment -n openshift-ingress
    NAME                       READY   UP-TO-DATE   AVAILABLE   AGE
    istiod-openshift-gateway   1/1     1            1           2m
    router-default             2/2     2            2           6h4m
    ```

> Gateway API and OSSM have been installed successfully and are ready to use. We are now ready to create Gateways and HTTPRoutes.

Reviewer 2: would it be a good idea to mention that at this point when all the deployments are successful, the gatewayClass will then show `Accepted` status?

In this section, we will show you how to install and configure Gateway API via the Ingress Operator. We will configure
a simple HTTPRoute and backend to demo basic Gateway API functionality.

> Note: Since this is still a Developer Preview, it is important to note that it may not be flawless and it is possible

Reviewer: This change is not grammatically correct.
```suggestion
Note: Since this is still a Developer Preview, it is important to note that it may not be flawless, and it is possible
```

for bugs to be present.

## Prerequisites

* A non-production OpenShift Cluster, version 4.13.5 or greater, on a cloud platform that supports Kubernetes service

>   load balancers (e.g. GCP, AWS and Azure)

Reviewer: We use the Oxford comma both by convention and by rule: https://github.com/openshift/enhancements/blob/master/guidelines/README.md?plain=1#L28
https://github.com/openshift/enhancements/blob/8d475dab4d176799814e1058264023a7f8762720/CONVENTIONS.md?plain=1#L58-L59
```suggestion
  load balancers (e.g. GCP, AWS, and Azure)
```

Author: Ack, good to know, I was looking at this with the heathen comma in mind

> * If OSSM is already installed, it must be OSSM 2.6.2

Reviewer: ```suggestion
* If OSSM is already installed, the minimum version must be OSSM 2.6.2
```

> gatewayclass.gateway.networking.k8s.io/openshift-default created

Reviewer 2: for us we always use the controller name as `controllerName: openshift.io/gateway-controller/v1`

Reviewer: This has to be updated to distinguish between old and new controllerNames.

```

Behind the scenes, the Ingress Operator creates a ServiceMeshControlPlane (SMCP) CR, which is the configuration API for OSSM. Wait a minute or two for the SMCP to be created and become ready:

```console
$ oc get smcp -n openshift-ingress
NAME                READY   STATUS            PROFILES      VERSION   AGE
openshift-gateway   5/5     ComponentsReady   ["default"]   2.4.0     10m
```

> **Version Note:** For OpenShift 4.20, the SMCP version will be 2.6.2 or greater.

Reviewer 2: if we create gatewayclass now, doesnt it automatically install OSSM 3?

```console
$ oc get smcp -n openshift-ingress
NAME                READY   STATUS            PROFILES      VERSION   AGE
openshift-gateway   5/5     ComponentsReady   ["default"]   2.4.0     107s
openshift-gateway   5/5     ComponentsReady   ["default"]   2.4.0     10m
```

OSSM will create two new Istio deployments in the `openshift-ingress` namespace. Ensure the pods for the deployments are running. These will need to become ready for the SMCP to also go ready:
```console

> $ oc get deployment -n openshift-ingress

Reviewer: I think you inadvertantly deleted this command.

    targetPort: 8080
```

First, create the Gateway with a listener (`spec.listeners` instance) that will listen on a port for HTTP requests that match a subdomain of the cluster domain. The Gateway must be created in the `openshift-ingress` namespace and specify the GatewayClass that we created earlier in the `spec.gatewayClassName` field.

> 

Reviewer 2: It might be worthwhile to mention that the gateway can be created in another namespace, but just wont have DNSRecords

Reviewer: I disagree.  Instead, let's keep it simple for those just getting started.

```

By default, Istio will [automatically provision](https://istio.io/latest/docs/tasks/traffic-management/ingress/gateway-api/#automated-deployment) a gateway deployment and service with the same name upon creation of this Gateway object:

> * **OpenShift 4.20 and beyond:** The deployment name changes to `demo-gateway-openshift-default`.

Reviewer 3: ```suggestion
* **OpenShift 4.19 and beyond:** The deployment name changes to `demo-gateway-openshift-default`.
```

    Application is not exposed. You can expose services to the outside world by executing one or more of the commands below:
     'oc expose service/foo-app'
    Run 'oc status' to view your app.
```
```
$ oc rollout status deployment -w -n demo-app foo-app

> oc rollout status deployment -w -n demo-app foo-app

Reviewer 3: Did you accidentally duplicate the line?

    Application is not exposed. You can expose services to the outside world by executing one or more of the commands below:
     'oc expose service/foo-app'
    Run 'oc status' to view your app.

> ```

Reviewer: This seems out of place.

EOF
httproute.gateway.networking.k8s.io/http created
```
Wait for the gateway deployment to be ready:
```console

> $ oc wait -n openshift-ingress --for=condition=ready gateways.gateway.networking.k8s.io demo-gateway

Reviewer 3: Would it make sense to keep this but change "Ready" to "Programmed"?

Reviewer: +1

Reviewer: Was this change inadvertantly deleted?

EOF
httproute.gateway.networking.k8s.io/http created
```

Wait for the gateway deployment to be ready:

  * **Prior to OpenShift 4.20:** The condition is `ready`.
    ```console
    $ oc wait -n openshift-ingress --for=condition=ready gateways.gateway.networking.k8s.io demo-gateway
    ```
  * **OpenShift 4.20 and beyond:** The condition is `programmed`.
    ```console
    $ oc wait -n openshift-ingress --for=condition=programmed gateways.gateway.networking.k8s.io demo-gateway
    gateway.gateway.networking.k8s.io/demo-gateway condition met
    ```

Now let’s send a request to the HTTPRoute we just created using an HTTP HEAD request to only get the headers. The `app` response header should have a value of "foo" for our demo application. You may have to wait a couple of minutes for the new DNS record to propagate before resolving:

```console
$ curl -I http://demo.gwapi.${DOMAIN}
```

You should see a 200 response code:

```
HTTP/1.1 200 OK
server: istio-envoy
app: foo
...
```

> Thanks for trying OpenShift Istio Gateway API. If you'd like to learn more, go to section on [Getting Creative](https://www.google.com/search?q=./GettingCreative.md).

Reviewer: It shouldn't need the new url.  If it does, use the correct github.com url.

EOF
httproute.gateway.networking.k8s.io/http created
```

Wait for the gateway deployment to be ready:

>   * **Prior to OpenShift 4.20:** The condition is `ready`.

Reviewer 3: ```suggestion
  * **Prior to OpenShift 4.20:** The condition is `Ready`.
```

>   * **Prior to OpenShift 4.20:** The condition is `Ready`.

Reviewer 3: Given my previous comments, I suspect this should be "4.19", not "4.20".  Please double-check.

    ```console
    $ oc wait -n openshift-ingress --for=condition=ready gateways.gateway.networking.k8s.io demo-gateway
    ```

>   * **OpenShift 4.20 and beyond:** The condition is `programmed`.

Reviewer 3: ```suggestion  
  * **OpenShift 4.20 and beyond:** The condition is `Programmed`.
```

>   * **OpenShift 4.20 and beyond:** The condition is `Programmed`.

Reviewer 3: Likewise, please double-check that "4.20" is correct.

    ```console
    $ oc wait -n openshift-ingress --for=condition=programmed gateways.gateway.networking.k8s.io demo-gateway
    gateway.gateway.networking.k8s.io/demo-gateway condition met
    ```

Now let’s send a request to the HTTPRoute we just created using an HTTP HEAD request to only get the headers. The `app` response header should have a value of "foo" for our demo application. You may have to wait a couple of minutes for the new DNS record to propagate before resolving:

```console
$ curl -I http://demo.gwapi.${DOMAIN}
```

You should see a 200 response code. The response will include headers from the application:

>   * **Prior to OpenShift 4.20:** Response includes `istio-envoy` server header.

Reviewer 3: ```suggestion
  * **Prior to OpenShift 4.19:** Response includes `istio-envoy` server header.
```
OpenShift 4.19 configures Istio not to set this header: https://github.com/openshift/cluster-ingress-operator/blob/release-4.19/pkg/operator/controller/gatewayclass/istio.go#L188-L191

    ```
    HTTP/1.1 200 OK
    server: istio-envoy
    app: foo
    ...
    ```

>   * **OpenShift 4.20 and beyond:** Response includes application server header and `app: foo`.

Reviewer 3: `app: foo` is in the previous example for OpenShift 4.19 and earlier too, and I believe it is set by the application, not Envoy.
